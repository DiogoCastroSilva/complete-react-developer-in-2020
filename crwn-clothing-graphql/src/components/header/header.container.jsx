import React from 'react';
import { Query } from 'react-apollo';
import { gql } from 'apollo-boost';

import { GET_CART_HIDDEN } from '../../grapql/resolvers'

import Header from './header.component';


const HeaderContainer = () => (
    <Query query={GET_CART_HIDDEN}>
        {
            ({ data }) => {
                const { cartHidden } = data;
                return <Header hidden={cartHidden} />
            }
        }
    </Query>
);

export default HeaderContainer;