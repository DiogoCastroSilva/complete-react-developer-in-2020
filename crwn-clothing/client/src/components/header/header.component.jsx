import React from 'react';

import {
    HeaderContainer,
    LogoContainer,
    OptionsContainer,
    OptionLink
} from './header.styles';

import { ReactComponent as Logo } from '../../assets/icons/crown.svg'
import CartItemContainer from '../../containers/cart-item.container';
import CartDropdownContainer from '../../containers/cart-dropdown.container';


const Header = ({ currentUser, hidden, signOutStart }) => (
    <HeaderContainer>
        <LogoContainer to='/'>
            <Logo />
        </LogoContainer>
        <OptionsContainer>
            <OptionLink
                className='option'
                to='/shop'
            >
                SHOP
            </OptionLink>
            <OptionLink
                className='option'
                to='/contact'
            >
                CONTACT
            </OptionLink>
            {currentUser ? 
                <OptionLink
                    as='div'
                    className='option'
                    onClick={signOutStart}
                >
                    SIGN OUT
                </OptionLink> :
                <OptionLink
                    className='option'
                    to='/signin'
                >
                    SIGN IN
                </OptionLink>    
            }
            <CartItemContainer />
        </OptionsContainer>
        {hidden ?
            null :
            <CartDropdownContainer />
        }
    </HeaderContainer>
);

export default Header;