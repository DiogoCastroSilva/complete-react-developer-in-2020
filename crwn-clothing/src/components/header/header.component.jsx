import React from 'react';
import { connect } from 'react-redux';

import {
    HeaderContainer,
    LogoContainer,
    OptionsContainer,
    OptionLink,
    OptionDiv
} from './header.styles';

import { ReactComponent as Logo } from '../../assets/icons/crown.svg'
import { auth } from '../../firebase/firebase.utils';
import CartIcon from '../cart-icon/cart-icon.component';
import CartDropdown from '../cart-dropdown/cart-dropdown.component';
import { selectCurrentUser } from '../../redux/user/user.selectors';
import { selectCartHidden } from '../../redux/cart/cart.selectors';
import { createStructuredSelector } from 'reselect';

const Header = ({ currentUser, hidden }) => (
    <HeaderContainer>
        <LogoContainer to='/'>
            <Logo />
        </LogoContainer>
        <OptionsContainer>
            <OptionLink
                className='option'
                to='/shop'
            >
                SHOP
            </OptionLink>
            <OptionLink
                className='option'
                to='/contact'
            >
                CONTACT
            </OptionLink>
            {currentUser ? 
                <OptionDiv
                    className='option'
                    onClick={() => auth.signOut()}
                >
                    SIGN OUT
                </OptionDiv> :
                <OptionLink
                    className='option'
                    to='/signin'
                >
                    SIGN IN
                </OptionLink>    
            }
            <CartIcon />
        </OptionsContainer>
        {hidden ?
            null :
            <CartDropdown />
        }
    </HeaderContainer>
);

const mapStateToProps = createStructuredSelector({
    currentUser: selectCurrentUser,
    hidden: selectCartHidden
});

export default connect(mapStateToProps)(Header);