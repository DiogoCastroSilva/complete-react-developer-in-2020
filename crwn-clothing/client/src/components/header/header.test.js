import React from 'react';
import { shallow } from 'enzyme';

import Header from './header.component';
import CartDropdownContainer from '../../containers/cart-dropdown.container';
import { OptionLink } from './header.styles';


describe('Testing Header component', () => {
    let wrapper;
    let handleSignOut;

    beforeEach(() => {
        handleSignOut = jest.fn();

        const mockProps = {
            currentUser: {
                uid: '123'
            },
            hidden: true,
            signOutStart: handleSignOut
        };

        wrapper = shallow(<Header {...mockProps} />);
    });
   
    it('should render Header component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe('if currentUser is present', () => {
        it('should render sign out link', () => {
            expect(
                wrapper
                    .find(OptionLink)
                    .at(2)
                    .text()
            ).toBe('SIGN OUT');
        });
    
        it('should call signOutStart method when link is clicked', () => {
            wrapper
                .find(OptionLink)
                .at(2)
                .simulate('click');
    
          expect(handleSignOut).toHaveBeenCalled();
        });
    });

    describe('if currentUser is null', () => {
        it('should render sign in link', () => {
            const mockProps = {
                hidden: true,
                currentUser: null,
                signOutStart: handleSignOut
            };
      
            const newWrapper = shallow(<Header {...mockProps} />);
      
            expect(
                newWrapper
                    .find(OptionLink)
                    .at(2)
                    .text()
            ).toBe('SIGN IN');
        });
    });

    describe('if hidden is true', () => {
        it('should not render CartDropdown', () => {
            expect(wrapper.exists(CartDropdownContainer)).toBe(false);
        });
    });
    
    describe('if currentUser is null', () => {
        it('should render CartDropdown', () => {
            const mockProps = {
                hidden: false,
                currentUser: null,
                signOutStart: handleSignOut
            };
    
            const newWrapper = shallow(<Header {...mockProps} />);
    
            expect(newWrapper.exists(CartDropdownContainer)).toBe(true);
        });
    });
});