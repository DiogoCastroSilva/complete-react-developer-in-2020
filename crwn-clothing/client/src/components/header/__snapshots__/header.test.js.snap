// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Header component should render Header component 1`] = `
<styled.div>
  <Styled(Link)
    to="/"
  >
    <ForwardRef(SvgCrown) />
  </Styled(Link)>
  <styled.div>
    <Styled(Link)
      className="option"
      to="/shop"
    >
      SHOP
    </Styled(Link)>
    <Styled(Link)
      className="option"
      to="/contact"
    >
      CONTACT
    </Styled(Link)>
    <Styled(Link)
      as="div"
      className="option"
      onClick={[MockFunction]}
    >
      SIGN OUT
    </Styled(Link)>
    <Connect(CartItemContainer) />
  </styled.div>
</styled.div>
`;
